### ISSUE TITLE
Aging Report

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#419]
In GitLab by @kdjstudios on Jul 12, 2017, 12:26

Amanda wrote: http://www.servicedesk.answernet.com/profiles/ticket/2017-07-12-75939/conversation

I ran an aging report and it looked normal, I closed billing and ran it again and it looks the same. Something is wrong! The current and 30 are combined together.. the columns did not move.

[aging__4__1056am_Report.pdf](/uploads/0d35175fd7f690b84318d7e37a55e6d3/aging__4__1056am_Report.pdf)

[aging__5__118pm_Report.pdf](/uploads/f57a1b61d00eefc6e506845b441e0572/aging__5__118pm_Report.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 17, 2017, 15:16

@smasih @tim.traylor 

Update request from Amanda:

Do we have an update on this.. I know Cheryl is about to send her aging's out for this week :)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 18, 2017, 24:46

@kdjstudios @tim.traylor 

I am sorry, I didn't understand the issue. What exactly is Amanda trying to tell..?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 18, 2017, 08:22

@smasih @tim.traylor 

Amanda is stating that the aging before and after she completed billing are the same. As the aging should show current balances before a billing cycle, and then reflect the new current balances after completing the billing. Thus the balances should be different, however in the files shown by amanda they are not.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 19, 2017, 02:34

@kdjstudios 

Thank's for the explanation and where as this query is concerned.... If you see closely at the first row **Acc: 188-SR5010** has bal **923.84** which means that no invoice was generated, otherwise  it should have affected the balance. Aging report considers **current date** as a parameter for comparison so it's obvious that report generated on the same day will give similar aging.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 19, 2017, 09:02

@tim.traylor 

I think we need to have this report run based upon  **current date and time**, because changes can be made on the same day as amanda just described, which according to @smasih is why this currently results in old information (I presume the calculation at midnight) be given rather then **current date and time**. 

Should we have this be marked as an current issue or one that needs to be added to the reporting overhaul?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 20, 2017, 01:24

@tim.traylor 

@kdjstudios that totally makes sense but, we can not filter invoices on the basis of seconds/time It has to be on the date only at max we can do it from 12 - to - 12  and that's the same thing.
The issue here, what I believe and actually is 
* the **amount**  which was same before and after billing.
* Invoice date was different than the generation date. Aging report as I mentioned before takes current date as parameter **so it matters when we are running the report however, the aging on the individual invoice should be absolutely accurate**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 20, 2017, 08:30

@smasih 

Does this report take into calculation of draft invoices? Because that would most certainly result in the same amounts. 

I will inform Amanda that the report only runs for days, and does not take into effect the time of day. @tim.traylor I still believe this should be based upon date and time, lets discuss this on our meeting.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 20, 2017, 10:11

@smasih @tim.traylor 

The team will being doing more research into why there are no changes in balances

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 24, 2017, 06:19

@kdjstudios @tim.traylor 

````ruby
agings = [0.00, 0.00, 0.00, 0.00, 0.00, 0.00]
today = Time.now().strftime("%Y-%m-%d").to_date # Mon, 24 Jul 2017
# This what comes in return Mon, 24 Jul 2017
aged_invoices = invoices.where(:balance.gt => 0.to_money)
agings[0] = aged_invoices.where(:invoice_date.gte => (today - 29.days)).sum(&:balance).to_money
````

So as we can see we are not considering time for aging calculation, and if we do it will make it overly complex. Lastly, Amanda would have already drafted/applied the billing, therefore, there was no change in aging report even after locking/finalizing the billing.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 10, 2017, 07:17

@smasih Hi Sumeet,  Sorry for the delay on this - it's been a rough week!

In a nutshell, when calculating aging against an invoice date we should NOT be including any amounts dated after the invoice date.  So for the situation we had reviewed where there was a payment dated 7/5/17 and the invoice date was 7/1/17, we should NOT be including the payment in the calculation.

I've attached the email where I detailed the situation to Gary and what his response was.  If you have any questions please let me know.

Thx, Tim[AgingDiscussion.pdf](/uploads/2f5708464079de82388805d90fe6760b/AgingDiscussion.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 17, 2017, 07:44

@tim.traylor 

Great, thanks for the resolution. The issue is with the **dynamic field Invoice balance** and aging code uses this field to calculate the balance so what we can do is add a boolean flag on payment to handle the invoice balance.

 We wil try implementing this approach tomorrow and if this works we will release it on this saturday.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 18, 2017, 09:24

@tim.traylor 

As we discussed yesterday over the call, to tackle the issue during the aging calculation, but during the implementation, I had following issues.  
* We Use balance to calculate the agings

````ruby
def calculate_agings(end_date = nil)
    agings = [0.00, 0.00, 0.00, 0.00, 0.00, 0.00]
    aged_invoices = invoices.where(:balance.gt => 0.to_money)
    today = Time.now().strftime("%Y-%m-%d").to_date
      
    agings[0] = aged_invoices.where(:invoice_date.gte => (today - 29.days)).sum(&:balance).to_money
    agings[1] = aged_invoices.where(:invoice_date => (today - 59.days)..(today - 30.days)).sum(&:balance).to_money
    agings[2] = aged_invoices.where(:invoice_date => (today - 89.days)..(today - 60.days)).sum(&:balance).to_money
    agings[3] = aged_invoices.where(:invoice_date => (today - 119.days)..(today - 90.days)).sum(&:balance).to_money
    agings[4] = aged_invoices.where(:invoice_date.lte  => (today - 120.days)).sum(&:balance).to_money
   return agings
end
````

* So, how to tell the Aging calculation that, *hey this is the payment I was talking about* .. there can be n existing payments with same or partial amount

Later I thought to make the Invoice balance calculation smarter during payment posting, but that lead to further errors because 
 * Invoice date is **Sat, 01 Jul 2017** but actual date is : **2017-07-05 16:06:31 UTC** 
 * Payment paid date is: **2017-07-05 00:00:00 UTC** whereas actual date is **2017-07-11 14:33:14 UTC**


**So, there are many other aspects which might lead to inaccurate calculations..... need more time to get this issue resolved**

changing priority to 0.5

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 18, 2017, 09:42

@smasih Hi Sumeet, I understand the problem.  Written as a kind of pseudo-code, what about this:

![image](/uploads/369934b40b4ff5df0f0c5c6058b8dd4f/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 22, 2017, 08:23

@tim.traylor @abakshi 

Unfortunately, We didn't have a call today but we very much wanted to discuss it with you. The process you have explained that was great but I think this will be more useful in curing existing invoices but for future, we can have additional balance field in invoice.

so, when a payment is posted it keeps the foreign key of Invoices it was paid against. Now from here we can make a check

if payment paid date is after invoice's official invoice date 
  add the paid amount to new field in Invoice
end

And while calculating the aging we can consider this field as well

and this field can be cleared of on next cycle.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 28, 2017, 06:30

@tim.traylor @abakshi 

Your's and our idea was same just, we were thinking to do that on new payment event, and yours is to fix the existing. 

````ruby
data=[]
account.payments.each do |payment|
	if payment.invoices.count > 0
		payment.invoices.each do |invoice|
			if payment.paid_date.to_time < invoice.created_at.to_time
				data.push(payment)
			end
		end
	end
end
````

This gives us the Array of payments we need to add back. But, we have hit a wall here ...
we are not saving the bifurcation of payment, due to which we are not able to pull amount that was distributed among the invoices. Some how even if we manage to do that ....  I am not able to think a way of calculating aging on that amount.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 29, 2017, 01:51

primary for #412

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Sep 13, 2017, 06:14

````ruby
  def apply_outstanding_invoice(invoice)
    if amount_left > invoice.balance
      #current_payment.amount = invoice.balance
      @amount_left -= invoice.balance
      invoice.balance = 0.to_money

      if invoice.invoice_date.to_time < current_payment.paid_date.to_time and invoice.created_at.to_time > current_payment.paid_date
          invoice.bal_4_same_day_payment =  amount_left
      else
          invoice.bal_4_same_day_payment = '0'.to_money
      end      
      invoice.save
    else
````
![Screenshot-3](/uploads/47184cf20ed2a952258c615c66c11afe/Screenshot-3.png)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 14, 2017, 07:46

Hi [Tim](@tim.traylor),

The Aging is a concept that was having multiple related issues in invoice calculation and invoice report reported by business (since we took over). Each Aging issue can be listed as a separate scenario for more in-depth analysis. 

We tried to resolve it by fixing a common root cause:
* Switching to use "Invoice Date" instead of "Current Date" during Aging calculation for individual invoice calculation and generation
* Data scrubbing to fix the existing data

Besides this, there were other specific issues that were fixed separately.

We understand that the above fix has neither created more issues nor it was incorrectly implemented, however, it may have limited effect and might not have resolved all the related underlying issues as it was intended or assumed when we marked it fixed.

We have further analyzed the specific scenario that was reported on July 25th call("same day payment issue"):

* Scenario
    * site: NCSM
    * account: Alliance Inspection Management - Ford Balance
    * account #: 001-2020

It seems that it might not be the "real issue" as per the latest understanding as on today. And here is the reason behind what made us think this way:

* **AIM Ford** had an Invoice dated in the past invoice date (**7/1**) and creation date as **7/5**. 

* A payment was made for **$163,419.32** on **7/11** with a payment date of **7/5**.

* So, from this, it's evident that invoice existed way before the payment was posted into the database.

* Payment mode was **Bank Wire** which means it was a manual payment. 

* We cannot post a manual payment through UI if there is an existing Invoice in draft stage for that account. 

* And, this makes clear that when the payment was posted, the invoice was in **locked/finalized** stage.

---

By providing detailed explanation above,  we do accept that there is a fault in invoice balance calculation and it's not being maintained properly. However, we do not understand what exactly might be causing it in order to fix it precisely. 

*With this new understanding (after our yesterday's call), we would like to have a discussion with you and reach a better conclusion.*

Few points that can be considered for further analysis:
* As per the current code, the payment only updates the balance for the respective invoice. (Let me reassure that there was no change in invoice aging for that particular invoice as it's already in locked stage).
* the First bug is that manually added items do not add up in invoice balance.
* We do not track the invoice lock/ finalization date so far, though it might have really helped in analyzing such issues.

**Comment by kdjstudios** [comment]
In GitLab by @abakshi on Sep 15, 2017, 04:57

Regarding the manual entry issue, @smasih identified that there is an issue with Firefox mozilla browser, where the manually entered line items do not add up on the invoice balance. But it works fine with Chrome. So this is the reason why Sumeet was unable to demonstrate the issue that he witnessed during the day.

Another issue in the similar scenario (root cause is Firefox mozila) is related to manual payments getting posted twice. This should not be confused with ticket #463 as this is a different issue, so we can add a separate ticket for this.

We can notify the users to use only Chrome till these issues are fixed.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 22, 2017, 07:12

@kdjstudios @tim.traylor 

Two major reason for this issue:-

* During the aging calculation of invoice in draft stage... It was considering all invoices except itself because it's still an instance but not created in db.  

* There was a "skip_all_callback" function which was executed in last during the lock/finalized stage, which is used to skip unnecessary work like recalculation of balance, reassignment of invoice number etc. Unfortunately, this was also affecting the aging calculation during lock. We moved aging calculation after this function. 


**Issue is fixed on merge, and I have tested on different accounts. Besides, I want someone from answernet to test that out as well and let us know if any discrepancy is found in individual aging and balance calculation.**

**It would be great if you guys can test it out today so that we can have an update on our Monday morning and release it _IF Aprooved_**

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 22, 2017, 09:44

* Testing can be done in merge, doesn't matter which account or site we pick as it is just a test server at the moment.
* All the email notifications on manual invoice generation and editing have been blocked. (This change is done for testing and is transient) 
* **So the invoice we generate might not be reflecting accurate aging balance, because it totally depends on the balance of previous invoices** 
* Any Invoice after this one has to be reflecting accurately aging values.
* We can generate manual invoices through account main page, but those will be included only in the current billing cycle.
* Best thing would be if we can run two or three billing process and see if invoice's aging buckets sum is equal to invoice total (importing the usage for any site as its test and no email notifications will be sent until you click **send email** button)
  * Please note we can run the same usage file over and over again because our motive is to check the balance trail and aging calculation on an individual invoice.
* We can also manually edit the invoice and post a credit on the invoice to check how the system is handling negative/credit invoices.
  * As credit amount in the Invoice Current section was also the reason of aging discrepancy earlier.
* We can post two or three test payments just to be certain that Invoice balance is being handled correctly.

* Lastly, today I will be up late so if there is anything I can help with, please feel free to text/call me.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 22, 2017, 13:03

Testing Process: Billings site ( I do not believe this site had any previous issues with their aging)
* Ran the export from the VCC Portal to generate the usage file. NOTE this file I used for each billing cycle again. [Billing_Export__17_.csv](/uploads/9064dbf752550e3abd9fc9f9b8c2f0d5/Billing_Export__17_.csv)
* First Billing cycle:
 * Started the cycle
 * accepted exceptions and applied billing
 * added stagged/additional fees to these accounts:
  * 071-1954 Ear, Nose & Throat Associates - Other: $100
  * 071-649 Project Management Balance - Other: $100
  * 071-784 Colonial Professional Ctr - Other: $300
 * finalize the cycle
* tried to the run the AR Snapshot report- only would go to the date 6/30/17? - [ar_snapshot__.csv](/uploads/39dc0a778468e775ca87ee553a9849e4/ar_snapshot__.csv)
* Ran the Aging report: [aging__.csv](/uploads/e3d4bfd578cac8973dc3572542461e05/aging__.csv)
* Ran the Aging by Invoice report: [aging_invoiced_10-25-2017_Master.csv](/uploads/30da838640989f0f3a5a87c8864c1982/aging_invoiced_10-25-2017_Master.csv)


**Already found an issue with the age current in the Aging report. all account values show "11,111,111,111". Stopping the testing until fixed. **

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 22, 2017, 16:25

Continuing Testing according to @smasih the result above is normal.

* Made payments to the following accounts:
  * 071-2283 Eye Physicians - Cash $100  
  * 071-3535 Polson Animal Clinic - Cash $100
  * 071-408 Exec Air Montana, Inc. - Cash $300
* Ran the following reports
  * Summary Billing Journal: [summary_billing_journal_10-25-2017_Master.csv](/uploads/4e2b7c2d569caef64c2960ced44644fe/summary_billing_journal_10-25-2017_Master.csv)
  * Ran the Aging Report: [aging____1_.csv](/uploads/afdc7ec83f1c3779fc05ba89ff1485e2/aging____1_.csv)
  * Ran the Aging by Invoice report: [aging_invoiced_10-25-2017_Master__1_.csv](/uploads/7495bd2c2fa2c254bfebe9cfd279fabb/aging_invoiced_10-25-2017_Master__1_.csv)
 
 *Current Balance seem to differ between all of the reports?*
 
* Second Billing cycle:
  * Started the cycle
  * accepted exceptions and applied billing
  * finalize the cycle
 
* Ran the Summary Billing Journal: [summary_billing_journal_11-25-2017_Master.csv](/uploads/9e2ff82be69f298b8c38141dce579bc0/summary_billing_journal_11-25-2017_Master.csv)
* Ran the Aging Invoiced: [aging_invoiced_11-25-2017_Master.csv](/uploads/61df4439bbb2b89689da9f7be33efcde/aging_invoiced_11-25-2017_Master.csv)
* Ran the Aging: [aging____2_.csv](/uploads/97c72716e061027a5bbbb0882fad7006/aging____2_.csv)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 25, 2017, 08:03

@kdjstudios 

Thanks for testing the Aging Out. 

"$1111111" in the Aging Report was my fault, I was just using a fixed figure for testing anyhow we should have removed it.

Our concern was individual invoice aging / Aging as Invoiced report at present. 

So, billing cycles you processed had their invoice balance calculated accurately. There was one account which was off in Billling MT and that was because of it's old balance.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 25, 2017, 08:37

@smasih @tim.traylor 

Would we not need to test this over the course of 30+ days to see if the aging does adjust correctly? Or is that something that can be done to simulate it has been 30+ days to insure calculations are updating correctly over time?

I feel the testing we have done was beneficial, but maybe this is my misunderstanding, yet I don't think it is actually testing aging correctly. As we did it all in one day and no time had passed, so no aging had occurred. Correct?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 26, 2017, 01:04

@kdjstudios 

Let's just bring that up in our today's call.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Sep 27, 2017, 06:07

@kdjstudios hi Kyle, doing multiple cycles will test the aging as it is calculated on the invoices.  Of the two aging reports, only the aging as invoiced can be tested this way.  The aging 'as of today' is based on the current date, so you are correct that it can't be tested the same way.  that said, if what's being calculated on the invoices is OK, I'm OK to move ahead with this.  Thx, Tim

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 27, 2017, 08:34

@smasih @tim.traylor 

I just checked three random accounts and it appears that the current fields and the aging row are being show/calculated correctly on the last two invoices. 

@smasih is there any way to check all accounts with a script or programmatically in the database? If so, lets confirm for all accounts. If not, I would approve this based upon the results I reviewed with three randomly selected accounts.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 27, 2017, 09:01

@kdjstudios 

Yes, we can do that but, please let us know about the site we ran the billing cycles for.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 27, 2017, 09:11

@smasih 

The Billings Site.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 28, 2017, 07:52

@kdjstudios 

Here is the [Aging as Invoiced Report](/uploads/3e3b4c084569ae4d524147b536d63d25/aging_as_invoiced_billings_mt.csv) for the Billings MT.

There is only one faulty account and that is because initial balance calculation was not right. We are using balance forward system which means the last balance will be transferred.

In order to rectify this **either** client has to pay in full (*which is obviously not possible*) **or** we got to manually update the last invoice.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 28, 2017, 08:55

@smasih @tim.traylor 

That is great news right!?

Tim - should we have the team push this out and then do the check for all site? Correcting the last invoice balance on any accounts that are found with incorrect balances?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 28, 2017, 09:11

@tim.traylor agrees that we should push this out to production and update anything that we know to be wrong.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 28, 2017, 12:40

@kdjstudios  @tim.traylor 

* Pushing on to production won't be an issue it would hardly take 1/2hr.
* But, It would be great if someone from the accounting team can have a look and run two or three billing cycles other than Billings MT (*on merge.sabilling.com*) 
* That will make us certain that changes we did are really working.
* Once that is done we can push it **whenever requested** even tomorrow if Tim agree and the testing goes well.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 28, 2017, 13:42

@smasih @tim.traylor 

I have provided Michelle McKee with the testing instructions and we will await for her feedback. I gave her access to the Columbus site on Merge and a usage file to upload for that site.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 29, 2017, 01:45

Perfect

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 2, 2017, 09:00

@smasih @tim.traylor 

Michelle McKee was able to test three billing cycles on the Columbus site and she did not see any errors. Here is her response below:

I ran three different billing cycles – 10/22/17, 11/22/17 & 12/22/17. I chose 5 different accounts and compared them month to month. The aging was correct on the all three billing cycles.

In addition to her checking; Please check programmatically all accounts for all three billing cycles.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 3, 2017, 24:15

@kdjstudios 

I also ran the **aging as invoiced report** for "Colombus OH" and found no aging discrepancy. I think we are ready to move the changes to production this Saturday if approved by @tim.traylor.

* We also need to start informing operations that we will be releasing the code on this weekend and they might get the invoices with unexpected numbers 
  **(** *for Eg: late fee calculation depends on balance and balance was getting waived off earlier and therefore there were not much of the late charges but now, there will late charges reflecting on the invoices* **)**

* Additionally, operations might receive calls from customers/clients as they are used to the old way of calculation.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 8, 2017, 08:30

@smasih @tim.traylor 

I have already sent out an email informing the operations to expect changes to the way lates fees were being charged. I will start to follow up on all of the aging tickets to see if things are resolved on Monday.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Oct 8, 2017, 23:07

Released on production.
